EIPS/eip-2315.md

```diff
@@ -13,55 +13,67 @@ requires: 3540, 3670, 4200
 
-Unlike these machines, the Ethereum Virtual Machine _does not_ provide subroutine operations.  Instead, they must be synthesized using the dynamic `JUMP` instruction, which takes its destination on the stack. Further, the EVM provides _only_ dynamic jumps, preventing the static analysis we need.
+### We need safe control-flow.
 
-### We want a safer control-flow facilty.
+The EVM has unusually high requirements for safety.  Not only do many smart contracts control inordinately large amounts of valuable Ether, but once placed on the blockchain any defects are visible to attackers and cannot be repaired.  In valid code is an  We need to statically validate important safety constraints code at initialization time.
 
-The EVM has unusually high requirements for safety.  Not only do many smart contracts control inordinately large amounts of valuable Ether, but once placed on the blockchain any defects are visible to attackers and cannot be repaired.  We propose here to statically validate, at initialization time, that code will not execute invalid instructions, jump to invalid locations, or underflow stack.  It will maintain consistent numbers of inputs and outputs for subroutines, and stack growth is bounded in the absence of recursion.
+### We need static control-flow.
 
-### We want an efficient control-flow facility.
+The EVM's lone dynamic jumps cause two major problems. First, the need to synthesize static jumps and subroutines with dynamic jumps uses space and gas with needlessly complex code, as we will show below.
 
-The presence of dynamic jumps and lack of subroutines and static jumps causes two major ineffiencies. First, the need to synthesize static jumps and subroutines with dynamic jumps wastes space and gas and causes needless complexity, as we will show below. 
+Worse, jumps that can dynamically branch to any destination in the code can cause quadratic "path explosions" when traversing the flow of control. For Ethereum, this is a denial-of-service vulnerability that prevents us, at initialization time, from validating the safe use of EVM code and from compiling EVM code to machine code.
 
-Worse, dynamic jumps can branch to any destination in the code, so quadratic "path explosions" are possible when traversing the flow of control. For Ethereum, this is a denial-of-service vulnerability that prevents us, at initialization time, from validating the safe use of EVM code and from compiling EVM code to machine code.
+### We need efficient control-flow.
+
+Efficient to write by hand, compile from high level labguages, validate at deploy time, interpret by VMs, and compile to machine code.
+
+Static jumps, conditional jumps, and subroutines are sufficient and efficient in space and time, as shown by historical experience and as we will show for the EVM below.
+
+**We need static jumps and subroutines and must deprecate dynamic jumps.**
 
 ## Specification
```

Comment:
[Peer Review] I happen to notice there is no mention of "fork block number". Is this intentional? What is the plan for rolling out? Does it depend on fork block number or does it depend on total difficulty or any other approach to enable the new opcodes and deprecate the old one?

Reply:
It can't be known until the fork is actually scheduled, and doesn't seem to me as part of the technical specification anyway.

Reply:
[Peer Review] I would suggest adding something like `As of FORK_BLOCK_NUMBER the implementation of EVM MUST support the following new opcodes and deprecate the following old opcodes`.

This follows many previous Core EIPs (e.g. eip-2718, eip-7, eip-140). It also bring the convenience when you need to discuss and describe scenarios of backward compatibility.